adal/src/main/java/com/microsoft/aad/adal/AcquireTokenRequest.java

```diff
@@ -610,13 +610,36 @@ void onActivityResult(final int requestCode, final int resultCode, final Intent
                     final String tenantId = data.getStringExtra(
                             AuthenticationConstants.Broker.ACCOUNT_USERINFO_TENANTID);
                     final UserInfo userinfo = UserInfo.getUserInfoFromBrokerResult(data.getExtras());
+
+                    // grab the fields tracked by x-ms-clitelem
+                    final String serverErrorCode = data.getStringExtra(AuthenticationConstants.Broker.CliTelemInfo.SERVER_ERROR);
+                    final String serverSubErrorCode = data.getStringExtra(AuthenticationConstants.Broker.CliTelemInfo.SERVER_SUBERROR);
+                    final String refreshTokenAge = data.getStringExtra(AuthenticationConstants.Broker.CliTelemInfo.RT_AGE);
+                    final String speRingInfo = data.getStringExtra(AuthenticationConstants.Broker.CliTelemInfo.SPE_RING);
+
+                    // create the broker AuthenticationResult
                     final AuthenticationResult brokerResult = new AuthenticationResult(accessToken, null,
                             expire, false, userinfo, tenantId, idtoken, null);
+
+                    // set the x-ms-clitelem fields on the result from the Broker
+                    brokerResult.setServerErrorCode(serverErrorCode);
+                    brokerResult.setServerSubErrorCode(serverSubErrorCode);
+                    brokerResult.setRefreshTokenAge(refreshTokenAge);
+                    brokerResult.setSpeRing(speRingInfo);
+
                     if (brokerResult.getAccessToken() != null) {
                         waitingRequest.getAPIEvent().setWasApiCallSuccessful(true, null);
                         waitingRequest.getAPIEvent().setCorrelationId(
                                 waitingRequest.getRequest().getCorrelationId().toString());
                         waitingRequest.getAPIEvent().setIdToken(brokerResult.getIdToken());
+
+                        // add the x-ms-clitelem info to the ApiEvent
+                        waitingRequest.getAPIEvent().setServerErrorCode(brokerResult.getServerErrorCode());
```

Comment:
It will be possible that new ADAL talking to old broker, then you won't get back all the spe ring related data. Is it ok to set null for the API event data?

Reply:
Ignore the above comment. I saw the code below that you're checking for null when setting the telemetry data.